client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/FullTest.java

```diff
@@ -38,63 +40,109 @@
+             var expected = Thread.currentThread().getContextClassLoader().getResourceAsStream("resources/pi0.txt");
+        ) {
+            assertThat(actual).hasSameContentAs(expected);
+        }
 
     }
 
     /*
-     * Finds the path of a download with the given resourceId by filtering the
-     * downloads
+     * This tests that the download inputstream can be consumed within the same
+     * Flowable. But, there is a problem with Java 11:
+     * https://bugs.openjdk.java.net/browse/JDK-8228970 You can view the test here :
+     * https://hg.openjdk.java.net/jdk/jdk/rev/e4cc5231ce2d
      */
-    private String findPath(final ISavedJobState state, final String resourceId) {
-        return state.getDownloads().stream()
-            .filter(dl -> dl.getResourceId().equals(resourceId))
-            .findFirst()
-            .map(dl -> dl.getProperties().get("path"))
-            .orElseThrow();
+    @Test
+    @Disabled
+    @SuppressWarnings("java:S1607")
+    void testWithDownloadInsideStream() throws Exception {
+
+        var properties = Map.of(
+            "service.palisade.port", "" + embeddedServer.getPort(),
+            "service.filteredResource.port", "" + embeddedServer.getPort());
+
+        var session = ClientManager.openSession("pal://dave@localhost/cluster", properties);
+        var query = session.createQuery("resource_id");
+        var publisher = query
+            .execute()
+            .thenApply(QueryResponse::stream)
+            .get();
+
+        Flowable.fromPublisher(FlowAdapters.toPublisher(publisher))
+            .filter(m -> m.getType() == MessageType.RESOURCE)
+            .map(Resource.class::cast)
+            .map(session::fetch)
+            .subscribe(new FlowableSubscriber<>() {
+
+                @Override
+                public void onNext(final Download t) {
+                    LOGGER.debug("## Got message: {}", t);
+                    try (var is = t.getInputStream()) {
+                        LOGGER.debug("## reading bytes");
+                        var ba = is.readAllBytes();
+                        LOGGER.debug("## read {} bytes", ba.length);
+                    } catch (Throwable e) {
+                        LOGGER.error("Got error reading input stream into byte array", e);
+                        throw new IllegalStateException("Got error reading input stream into byte array", e);
+                    }
+                }
+
+                @Override
+                public void onError(final Throwable t) {
+                    LOGGER.error("## Error: {}", t.getMessage());
+                    Assertions.fail("Failed due to:" + t.getMessage());
+                }
+
+                @Override
+                public void onComplete() {
+                    LOGGER.debug("## complete");
+
+                }
+
+                @Override
+                public void onSubscribe(final org.reactivestreams.@NonNull Subscription s) {
+                    s.request(Long.MAX_VALUE);
```

Comment:
I know this is a test, but...
Is there a util for creating a `forEach`-type subscriber?

Reply:
I'll create one. This is the test that failed before.

Reply:
I've had another go at this with multiple JDKs. It fails when consuming the download in the stream. This test fails when enabled with a known JDK bug. We may have to make the client JDK 14 only?

Reply:
You can convert the publisher into any of the reactive libraries and use their for-each methods.